### PR TITLE
Implement LoadBalancerSourceRanges

### PR DESCRIPTION
**Pull request by danwinship**
untested, and probably depends on https://github.com/kubernetes/kubernetes/pull/125219

(includes #79)

### CODE REVIEW

**Comment by aojea** [comment]
https://github.com/kubernetes/kubernetes/pull/125219  merged, let's run it against it

**Comment by aojea** [comment]
we may have a bug somewhere https://github.com/kubernetes-sigs/cloud-provider-kind/issues/71, the status gets wiped out,
```
  I0530 21:28:19.605960 73666 loadbalancer.go:359] Service Status: {{[{172.18.0.6  0xc001563510 [{80 UDP <nil>}]}]} []}
  STEP: changing the UDP service's port @ 05/30/24 21:28:19.605
  I0530 21:28:19.658081 73666 loadbalancer.go:373] Service Status: {{[]} []}
  [PANICKED] in [It] - runtime/panic.go:114 @ 05/30/24 21:28:19.658
  I0530 21:28:19.658624 73666 util.go:81] 
```

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cloud-provider-kind/80/pull-cloud-provider-kind-conformance-parallel-ga-only/1796286411100393472

**Comment by danwinship** [comment]
OK, well, the ccm logs show that it saw the source ranges and did the "right" thing with them:

```
I0530 21:22:23.718309    6065 proxy.go:276] updating loadbalancer with config 
resources:
- "@type": type.googleapis.com/envoy.config.listener.v3.Listener
  name: listener_IPv4_80_TCP
  address:
    socket_address:
      address: 0.0.0.0
      port_value: 80
      protocol: TCP
  filter_chains:
  - filters:
    - filter_chain_match:
        source_prefix_ranges:
        - "10.244.2.16/32"
        - "172.18.0.3/32"
    - name: envoy.filters.network.tcp_proxy
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy
        access_log:
        - name: envoy.file_access_log
          typed_config:
            "@type": type.googleapis.com/envoy.extensions.access_loggers.stream.v3.StdoutAccessLog
        stat_prefix: tcp_proxy
        cluster: cluster_IPv4_80_TCP
```

but it didn't work. So I guessed how to use the API incorrectly. Maybe someone with more Envoy clue could look at this?

**Comment by aojea** [comment]
> but it didn't work. So I guessed how to use the API incorrectly. Maybe someone with more Envoy clue could look at this?

@howardjohn can you help us with the config needed in envoy to filter some specific ranges ?

File: pkg/loadbalancer/proxy.go [hunk 1]
```diff
@@ -234,6 +249,12 @@ func generateConfig(service *v1.Service, nodes []*v1.Node) *proxyConfigData {
 		}
 	}
 	lbConfig.ServicePorts = servicePortConfig
+
+	// LBSR validation allows whitespace :-/
+	for _, sr := range service.Spec.LoadBalancerSourceRanges {
+		lbConfig.SourceRanges = append(lbConfig.SourceRanges, strings.TrimSpace(sr))
+	}
+
 	klog.V(2).Infof("haproxy config info: %+v", lbConfig)
```
Comment (howardjohn) [inline pkg/loadbalancer/proxy.go:258]: It's envoy now :slightly_smiling_face:

**Comment by howardjohn** [comment]
I think you might need something like

```
            source_prefix_ranges:
            - address_prefix: 1.2.3.4
              prefix_len: 32
```

But that seems like it should give a pretty loud error

**Comment by danwinship** [comment]
> But that seems like it should give a pretty loud error

The envoy logs don't appear to end up in the e2e artifacts so I guess it wouldn't matter how loud it was...

**Comment by aojea** [comment]
/test pull-cloud-provider-kind-conformance-parallel-ga-only


can you rebase, it should be possible to see the logs now

**Comment by aojea** [comment]
finding the logs is still complex, I will change the logging to use the loadbalancer name namespace instead of the hash, but right now is failing with

```
2024-06-04 12:07:47.303][1][info][upstream] [source/common/upstream/cds_api_helper.cc:71] cds: added/updated 1 cluster(s), skipped 0 unmodified cluster(s)
[2024-06-04 12:07:47.312][1][warning][config] [source/common/protobuf/message_validator_impl.cc:61] Unknown field: type envoy.service.discovery.v3.DiscoveryResponse reason INVALID_ARGUMENT: invalid JSON in  envoy.service.discovery.v3.DiscoveryResponse @  resources[0].<any>.filter_chains[0].filters[0]: message envoy.config.listener.v3.Filter,   near 1:67   (offset 66):   no such   field: 'filter_chain_match'
[2024-06-04 12:07:47.312][1][warning][config] [source/extensions/config_subscription/filesystem/filesystem_subscription_impl.cc:63] Filesystem config update rejected: Proto constraint validation failed (ListenerValidationError.FilterChains[0]: embedded message failed validation | caused by FilterChainValidationError.Filters[0]: embedded message failed validation | caused by FilterValidationError.Name: value length must be at least 1 characters): name: "listener_IPv4_80_TCP"
address {
  socket_address {
    address: "0.0.0.0"
    port_value: 80
  }
}
filter_chains {
  filters {
  }
  filters {
    name: "envoy.filters.network.tcp_proxy"
    typed_config {
      [type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy] {
        stat_prefix: "tcp_proxy"
        cluster: "cluster_IPv4_80_TCP"
        access_log {
          name: "envoy.file_access_log"
          typed_config {
            [type.googleapis.com/envoy.extensions.access_loggers.stream.v3.StdoutAccessLog] {
            }
          }
        }
      }
    }
  }
}
```

**Comment by danwinship** [comment]
maybe fixed, but only for TCP because I couldn't figure out for sure where to add it for UDP

**Comment by aojea** [comment]
> maybe fixed, but only for TCP because I couldn't figure out for sure where to add it for UDP

good enough, we log an issue just to remember, but at this pace we have better loadbalance features that most cloud providers XD

/lgtm
/approve

**Pull request merged**